internal/getproviders/package_location_local_archive.go

```diff
@@ -56,13 +58,26 @@ func (p PackageLocalArchive) InstallProviderPackage(_ context.Context, meta Pack
 
 	filename := meta.Location.String()
 
-	// NOTE: We're not checking whether there's already a directory at
-	// targetDir with some files in it. Packages are supposed to be immutable
-	// and therefore we'll just be overwriting all of the existing files with
-	// their same contents unless something unusual is happening. If something
-	// unusual _is_ happening then this will produce something that doesn't
-	// match the allowed hashes and so our caller should catch that after
-	// we return if so.
+	// NOTE: Packages are immutable, but we may want to skip overwriting the existing
+	// files in due to specific scenarios defined below.
+
+	if entries, err := os.ReadDir(targetDir); err == nil && len(entries) != 0 {
+		// If the package might already be installed, we should try to skip overwriting the contents.
+		// When run with TF_PLUGIN_CACHE_DIR or similar, a given provider might already be executing
+		// and therefore locking the provider binary in the target directory (preventing the overwrite below)
+		//
+		// This does incur the overhead of two additional hash computations and could be
+		// skipped with smarter checks around re-use scenarios in the future.
+
+		targetHash, targetErr := PackageHashV1(PackageLocalDir(targetDir))
+		fileHash, fileErr := PackageHashV1(meta.Location)
```

Comment:
It's kinda annoying that we will already have calculated the same `fileHash` indirectly above by calling `meta.MatchesAnyHash` but the result is discarded before that function returns. :confounded: 

Not a huge deal -- the OS filesystem cache will probably save us most of the time -- but maybe one day (not in this PR) we can find a better way to structure this so that we don't end up needing to repeat this work.